Automattic/pocket-casts-android
# Advanced settings menu

## Description

Advanced settings menu to house various tweaks that might not be mainstream for users. This helps keep them out of the way. Metrics to these options will be added to determine if they warrant a "graduation" down the road. 

Advanced options will always default to a state that does not change the default behavior of the app.

The advanced settings in this PR add functionality that allows the user to disable sync while on a metered network. This prevents podcasts from getting added to the "up-next" queue, which would cause streaming activity to occur for podcasts added to the top of the list. By performing sync on only unmetered networks, a user can generally guarantee that if they configure podcasts to be automatically added to the up-next queue and downloaded when added to the up-next queue, all podcasts in the up-next queue will be fully downloaded and not updated automatically until they return to an unmetered network.

Fixes #839 

## Testing Instructions

Automated testing has been added at the unit and instrumented level. Instrumented tests include verifying the default settings / behavior before the advanced options were added and the effects it has on settings that derive a behavior from these settings.

For manual testing the main thing to observe is:

- When "Sync on Metered" is enabled **_(default)_**
    - No changes from prior releases 
- When "Sync on Metered" is disabled 
    - While connected to wifi **(unmetered)**, there is no change in sync behavior
    - While connected to a metered network:
        - Sync will not happen automatically when you open the app. You'll notice the "Last refresh" on the profile page will not update the time
        - Sync will not happen in the background if you have _Settings -> Storage & Data Use -> Background Refresh_ **(enabled)** 




Background sync is enabled by default. To test this PR:

1. Tap on the Profile tab
2. Note that "Last refreshed" is close to 0
3. Tap on the gear in the upper right
4. Tap on Advanced
5. Tap on "sync on Metered Networks" to disable 
6. Close the app completely
7. Open the app back up a minute or so later 
8. Tap on the Profile tab
9. Observe the Last refresh is not 0
10. Tap Refresh now
11. Observe the refresh occurs

## Screenshots or Screencast 

![image](https://user-images.githubusercontent.com/12544461/232259935-2b617676-dc9e-4a13-8843-44a1141f632c.png)
![image](https://user-images.githubusercontent.com/12544461/234450668-f5df08b1-1651-4036-9891-aca678d5b703.png)


https://github.com/Automattic/pocket-casts-android/pull/885#discussion_r1170761086



## Current Development Tasks

- [x] Add [advanced preferences menu](https://github.com/Automattic/pocket-casts-android/issues/858#issuecomment-1506912325) 
- [x] verify preference menu is wired in correctly to the menu
- [x] Add logic to support #839
    - [x] Add unit/instrumented tests to verify settings and expectations
    - [x] Ensure the options are disabled when background sync is disabled (I'm going to just hide it and replacing it with a SettingsRow with primary text. Couldn't figure how I'd do a grey-out yet)
    - [x] manual testing of app open on and off metered network 
    - [x] See if there's other workmanager related settings that warrant separate tweakable options. (See [comment](https://github.com/Automattic/pocket-casts-android/pull/885#issuecomment-1509958157))
- [x] Re-evaluate #858 to see if it warrants the extra work as the implementation requires some INTENT mechanics that WorkManager simplifies
- [x] Search the backlog of issues that might fall under the category of advanced and see if there's any interesting / useful settings to include as well.
- [x] Sanity check any method naming and settings to make sure the meaning is clear.
- [x] Update Advanced Page with Possibly something like in this [review](https://github.com/Automattic/pocket-casts-android/pull/885#discussion_r1170279509)
- [x] Walk around with these advanced options for a week and dogfood them.
 

## Checklist



#### I have tested any UI changes...



For the last bullet, I modeled everything after an existing compose fragment and page. Not sure if there's anything to be done there.
